### ISSUE TITLE
set pmix.locstr and pmix.cpuset

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-pmix#31]
Problem: ompi asks for `pmix.locstr` and `pmix.cpuset` (optionally).

Furthermore, `pmix.locstr` is listed in the business card exchange use case in  spec v5 sec B.1.2.

I think if we provide these attributes, then every rank of ompi won't need to gather hwloc info from scratch, as an `strace` would seem to indicate they are doing now.

**Comment by garlick** [comment]
Notes to whoever ends up doing this:

The two keys metioned above are referred to in ORTE code using these names
```
$ git grep pmix.locstr
include/pmix_common.h.in:#define PMIX_LOCALITY_STRING                "pmix.locstr"           // (char*) string describing a proc's location
$ git grep pmix.cpuset
include/pmix_common.h.in:#define PMIX_CPUSET                         "pmix.cpuset"           // (char*) hwloc bitmap applied to proc upon launch
```
Don't go looking at slurm for a reference since it apparently does this wrong. 

References on this from Ralph in #58 are:
- [pmix-standard](https://github.com/pmix/pmix-standard/blob/master/Chap_API_Server.tex#L308-518)
- [orte](https://github.com/openpmix/prrte/blob/master/src/prted/pmix/pmix_server_register_fns.c) (yeah the 700 line for loop)

For better or worse, in flux-pmix I didn't sue the preprocessor macros in the "standard" pmix.h to manipulate the info vectors.  Instead the "infovec" class in flux-pmix does that stuff.  That may be confusing when looking at orte code since it uses the macros.

Be prepared for some studying of the above pmix-standard, the orte code, and [flux-pmix](https://github.com/flux-framework/flux-pmix/blob/main/src/shell/plugins/main.c#L285) and some head scratching as there seem to be many ways to organize the "info schema".  Do you set the keys once per node, once for the whole system, once per task...?

It may be helpful to look at the code in openmpi to see how it is looking up the keys and do it that way.

**Comment by garlick** [comment]
FWIW my  Ubuntu 22.04 systemd.service(5) says:
> It is generally recommended to use Type=simple for long-running services whenever possible, as it is the simplest and fastest option.

**Comment by grondo** [comment]
A kludgy workaround that I hesitate to even mention would be to dump the shell's XML to a file and set `HWLOC_XMLFILE` for whatever process is going to needlessly call `hwloc_topology_load()`. It would be better to do this the right way, but the right way in this case sounds.. painful?

**Comment by garlick** [comment]
> A kludgy workaround that I hesitate to even mention would be to dump the shell's XML to a file and set HWLOC_XMLFILE for whatever process is going to needlessly call hwloc_topology_load().

Yeah we should do it right, but since openmpi may not be the only application-level code that does this, a shell option like `-o hwloc=xmlfile` or similar might be a useful thing to have around...

**Comment by wihobbs** [comment]
I agree it'd be useful, but we'd have to be sure that users retain the power to subsequently `unset` that variable, because it seems to cause mvapich jobs to fail:
```
(s=1,d=1) corona188 ~/scale-testing $ flux run -n48 ./hello-mvapich-corona
f2cFjYqSX: completed MPI_Init in 0.476s.  There are 48 tasks
f2cFjYqSX: completed first barrier in 0.000s
f2cFjYqSX: completed MPI_Finalize in 0.004s
(s=1,d=1) corona188 ~/scale-testing $ export HWLOC_XMLFILE=$(hostname).xml
(s=1,d=1) corona188 ~/scale-testing $ flux run -n48 ./hello-mvapich-corona
[corona188:mpi_rank_0][smpi_load_hwloc_topology] Please retry after setting MV2_BCAST_HWLOC_TOPOLOGY=0
Warning: mv2_get_arch_type: Failed to determine number of processors.
Warning: mv2_get_arch_type: Failed to determine number of processors.
```

**Comment by grondo** [comment]
Yes, I think the idea is that the hwloc XML would only be saved to `HWLOC_XMLFILE` if `-o pmi=pmix` is set (and thus this plugin is activated), or a separate `-o hwloc-xmlfile` or similar shell option was used (so it is opt-in only). 

One nice thing about doing it in the flux-pmix plugin is that once this issue is closed, the automated use of `HWLOC_XMLFILE` could be removed, so users don't have to worry about what plugin version they're running and do they also need to specify some other shell option to get good performance.

**Comment by grondo** [comment]
Though @garlick makes a good point about other use cases for `-o hwloc=xmlfile`, so maybe we should add that option to the job shell, and some way to activate it from other shell plugins.

**Comment by garlick** [comment]
Oops this should remain open for "correct" solution.
